t2i pipeline (#3932)
huggingface/diffusers: docs/source/en/api/pipelines/stable_diffusion/adapter.mdx

![img](https://huggingface.co/datasets/diffusers/docs-images/resolve/main/t2i-adapter/color_palette.png)

Next, create the adapter pipeline

```py
import torch
from diffusers import StableDiffusionAdapterPipeline, T2IAdapter

adapter = T2IAdapter.from_pretrained("diffusers/t2iadapter_color_sd14v1")
pipe = StableDiffusionAdapterPipeline.from_pretrained(
    "CompVis/stable-diffusion-v1-4",
    adapter=adapter,
    torch_dtype=torch.float16,
)
pipe.to("cuda")
```

Finally, pass the prompt and control image to the pipeline

```py
# fix the random seed, so you will get the same result as the example
generator = torch.manual_seed(7)

out_image = pipe(
    "At night, glowing cubes in front of the beach",
    image=color_palette,
    generator=generator,
).images[0]
```

![img](https://huggingface.co/datasets/diffusers/docs-images/resolve/main/t2i-adapter/color_output.png)

## Available checkpoints

Non-diffusers checkpoints can be found under [TencentARC/T2I-Adapter](https://huggingface.co/TencentARC/T2I-Adapter/tree/main/models).

### T2I-Adapter with Stable Diffusion 1.4

| Model Name | Control Image Overview| Control Image Example | Generated Image Example |
|---|---|---|---|

> |[diffusers/t2iadapter_color_sd14v1](https://huggingface.co/diffusers/t2iadapter_color_sd14v1)<br/> *Trained with spatial color palette* | A image with 8x8 color palette.|<a href="https://huggingface.co/datasets/diffusers/docs-images/resolve/main/t2i-adapter/color_sample_input.png"><img width="64" style="margin:0;padding:0;" src="https://huggingface.co/datasets/diffusers/docs-images/resolve/main/t2i-adapter/color_sample_input.png"/></a>|<a href="https://huggingface.co/datasets/diffusers/docs-images/resolve/main/t2i-adapter/color_sample_output.png"><img width="64" src="https://huggingface.co/datasets/diffusers/docs-images/resolve/main/t2i-adapter/color_sample_output.png"/></a>|

Reviewer: Let's maybe add this to https://huggingface.co/TencentARC

Reviewer 2: The checkpoints are in https://huggingface.co/diffusers. Once the model cards are added we can bulk transfer to the org.

The abstract of the paper is the following:

*The incredible generative ability of large-scale text-to-image (T2I) models has demonstrated strong power of learning complex structures and meaningful semantics. However, relying solely on text prompts cannot fully take advantage of the knowledge learned by the model, especially when flexible and accurate structure control is needed. In this paper, we aim to ``dig out" the capabilities that T2I models have implicitly learned, and then explicitly use them to control the generation more granularly. Specifically, we propose to learn simple and small T2I-Adapters to align internal knowledge in T2I models with external control signals, while freezing the original large T2I models. In this way, we can train various adapters according to different conditions, and achieve rich control and editing effects. Further, the proposed T2I-Adapters have attractive properties of practical value, such as composability and generalization ability. Extensive experiments demonstrate that our T2I-Adapter has promising generation quality and a wide range of applications.*

> 

Reviewer 2: Would make sense to add `HimariO` as a co-contributor to the pipeline if we're using their stuff.

Author: done!

## Available Pipelines:

| Pipeline | Tasks | Demo
|---|---|:---:|
| [StableDiffusionAdapterPipeline](https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_adapter.py) | *Text-to-Image Generation with T2I-Adapter Conditioning* | -

## Usage example

In the following we give a simple example of how to use a *T2IAdapter* checkpoint with Diffusers for inference.
The inference pipeline is the same for all pipelines:

 1. Take an image and run it through a pre-conditioning processor to obtain *control image*.
 2. Run the pre-processed *control image* and *prompt* through the [`StableDiffusionAdapterPipeline`].

Let's have a look at a simple example using the [Color Adapter](https://huggingface.co/RzZ/sd-v1-4-adapter-color).

```python
from diffusers.utils import load_image

image = load_image("https://huggingface.co/RzZ/sd-v1-4-adapter-color/resolve/main/color_ref.png")
```

![img](https://huggingface.co/RzZ/sd-v1-4-adapter-color/resolve/main/color_ref.png)

> Then we can create our color palette by simply resize it to 8 by 8 pixels then scale it back to original size.

Reviewer 2: ```suggestion
Then we can create our color palette by simply resizing it to 8 by 8 pixels and then scaling it back to original size.
```

```python
from PIL import Image

color_palette = image.resize((8, 8))

> color_palette = color_palette.resize((512, 512), resample=Image.Resampling.NEAREST)

Reviewer 2: What is the advantage of having to perform two resizing? Why not directly resize to 512x512?

Author: I wondered the same but this seems actually quite clever. The resize down creates the color pallete via however the resize algorithm aggregates pixel colors

```

Let's take a look at the processed image.

![img](https://huggingface.co/RzZ/sd-v1-4-adapter-color/resolve/main/color_palette.png)

> After we having `color_palette` in hand, we can create the [`StableDiffusionAdapterPipeline`] with pretrained checkpoint.

Reviewer 2: ```suggestion
After having `color_palette` in hand, we can create the [`StableDiffusionAdapterPipeline`] with a pretrained checkpoint.
```

```py
import torch
from diffusers import StableDiffusionAdapterPipeline, T2IAdapter

adapter = T2IAdapter.from_pretrained("RzZ/sd-v1-4-adapter-color")
pipe = StableDiffusionAdapterPipeline.from_pretrained(
    "CompVis/stable-diffusion-v1-4",
    adapter=adapter,
    torch_dtype=torch.float16,
)
pipe.to("cuda")
```

> And finally we feed the data to the pipelien and wait for the result!

Reviewer 2: ```suggestion
And finally we feed the data to the pipeline and wait for the result!
```

```py
# fix the random seed, so you will get the same result as the example
generator = torch.manual_seed(7)

out_image = pipe(
    ["At night, glowing cubes in front of the beach"],
    image=[color_palette],
    generator=generator,

> ).images[0]

Reviewer 2: Do we always have to pack these inputs as lists? For single inputs, I would imagine having them as lists. But okay for me.

Author: removed!

```

This should take only few seconds on GPU (depending on hardware). The output image then looks as follows:

> ![img](https://huggingface.co/RzZ/sd-v1-4-adapter-color/resolve/main/color_output.png)

Reviewer 2: Awesome stuff 😱

|[RzZ/sd-v1.4-adapter-seg](https://huggingface.co/RzZ/sd-v1-4-adapter-seg)<br/>*Trained with semantic segmentation*  | An [custom](https://github.com/TencentARC/T2I-Adapter/discussions/25) segmentation protocol image.|<a href="https://huggingface.co/RzZ/sd-v1-4-adapter-seg/resolve/main/sample_input.png"><img width="64" src="https://huggingface.co/RzZ/sd-v1-4-adapter-seg/resolve/main/sample_input.png"/></a>|<a href="https://huggingface.co/RzZ/sd-v1-4-adapter-seg/resolve/main/sample_output.png"><img width="64" src="https://huggingface.co/RzZ/sd-v1-4-adapter-seg/resolve/main/sample_output.png"/></a> |

## Mix and match multiple adapters

[`StableDiffusionAdapterPipeline`] also support using multiple type of *control image* at once with combination with [`MultiAdapter`].

> Here is a example of using keypose adapter for character posture control and depth adapter for outlining background.

Reviewer 2: ```suggestion
Here is an example of using keypose adapter for character posture control and depth adapter for outlining background.
```

Amazing stuff this!

Just like the previous example, we will first prepare the *control image* for inference. One big difference when using [`MultiAdapter`] is that the *control image* we will send to pipeline is 
combined from multiple images. In this example we stack two 3 channels RGB image(`cond_keypose`, `cond_depth`) together to create a 6 channels image tensor(`cond`).

```py
import torch
from PIL import Image
from diffusers.utils import load_image

cond_keypose = load_image(
    "https://huggingface.co/RzZ/sd-v1-4-adapter-keypose-depth/resolve/main/sample_input_keypose.png"
)
cond_depth = load_image("https://huggingface.co/RzZ/sd-v1-4-adapter-keypose-depth/resolve/main/sample_input_depth.png")
cond = [[cond_keypose, cond_depth]]

prompt = ["A man waling in an office room with nice view"]
```

Two *control image* should look like follows:

![img](https://huggingface.co/RzZ/sd-v1-4-adapter-keypose-depth/resolve/main/sample_input_keypose.png)
![img](https://huggingface.co/RzZ/sd-v1-4-adapter-keypose-depth/resolve/main/sample_input_depth.png)

> Now we can using `from_adapters` method combine keypose and depth adapter into one, then pass our newly created [`MultiAdapter`] to 

Reviewer 2: We're not using this method in the below code snippet it seems.

Author: removed!

[`StableDiffusionAdapterPipeline`]. You can also play around the value of `adapter_conditioning_scale` to balance the control between adapters.

```py
from diffusers import StableDiffusionAdapterPipeline, MultiAdapter

adapters = MultiAdapter(
    [
        T2IAdapter.from_pretrained("RzZ/sd-v1-4-adapter-keypose"),
        T2IAdapter.from_pretrained("RzZ/sd-v1-4-adapter-depth"),
    ]
)
adapters = adapters.to(torch.float16)

pipe = StableDiffusionAdapterPipeline.from_pretrained(
    "CompVis/stable-diffusion-v1-4",
    torch_dtype=torch.float16,
    adapter=adapters,
)

images = pipe(prompt, cond, adapter_conditioning_scale=[0.8, 0.8])
```

After prompt and image is processed by pipeline we should get the result looks like:

![img](https://huggingface.co/RzZ/sd-v1-4-adapter-keypose-depth/resolve/main/sample_output.png)

## T2I Adapter vs ControlNet

> T2I-Adapter is similar to ControlNet. However, T2i-Adapter uses a smaller auxiliary network which is only run once for the entire diffusion process. T2I-Adapter performs slightly worse than ControlNet. However, T2I-Adapter is cheaper to run and is cheaper to run multiple auxiliary networks.

Reviewer 2: ```suggestion
T2I-Adapter is similar to [ControlNet](https://huggingface.co/docs/diffusers/main/en/api/pipelines/controlnet). However, T2i-Adapter uses a smaller auxiliary network which is only run once for the entire diffusion process. T2I-Adapter performs slightly worse than ControlNet. However, T2I-Adapter is cheaper to run. It is also cheaper to run when using multiple auxiliary networks.
```
